Automattic/woocommerce-payments
# UPE reads customer details from checkout fields

Fixes #1981 

#### Changes proposed in this Pull Request


# UPE Reads Customer Details from Checkout Fields

> As noted in this Github thread, the customer details are not displayed when checking out with UPE. The most likely issue is that the customer details are not being properly added to the UPE transaction.
> ...
> We have gotten the latest feature description from Stripe for how to implement the additional billing details and hide the appropriate fields from UPE to avoid the customer having to enter information twice. We can now hide the fields that are duplicated in the billing details from UPE and instead provide the details from the billing details when confirming the payment. The fields in particular that should be hidden are: name, country, and postal code.
>
> In addition to the fields to hide, we should also provide as many of the billing fields as possible when confirming the intent. In particular: name, email, phone, and address fields.

This PR allows checkout fields to be read by the UPE upon the point of payment confirmation, so that we can add the customer's details to the payment transaction details. This should only affect the UPE when it is loaded on the checkout page: in other pages, such as the Order Pay page or the Add Payment Method page, there are no customer details present, so these pages will continue to work as they have been.

Once concern that I noticed is that even though we tell Stripe to hide certain fields when we mount the UPE, I noticed that with Sofort the country field remained present inside the UPE. If I remove the country argument from the `payment_method_data` billing details when I confirm the payment, I receive this error.

> You specified "never" for fields.billing_details.address.country when creating the payment Element, but did not pass confirmParams.payment_method_data.billing_details.address.country when calling stripe.confirmPayment or stripe.confirmSetup. If you opt out of collecting data via the payment Element using the fields option, the data must be passed at confirm-time.

This implies that I'm setting the flag for Stripe to hide this field correctly, but for whatever reason, it doesn't seem to hide the field, so I'm not entirely clear what's going on here. 

Everything else seems to be working smoothly enough.




#### Testing instructions


1. Ensure you have pulled the changes in this PR and the latest version of the Payments Server. Also, please enable the UPE gateway using the WCPay Dev Plugin or simply set _wcpay_feature_upe to a value of 1 in your wp_options table. 
2. Add a product to your cart and proceed to the checkout screen. 
3. Complete the checkout process as usual.
4. Either go to **Payments > Transactions** and select the completed transaction or view the completed order and click on the payment intent ID near the top of the page to be transported to the details page for this latest transaction.
![Payment intent ID on order page](https://cdn-std.droplr.net/files/acc_1104206/5nfx4d)
_Payment intent ID on order page_

5. Confirm that customer details from checkout are present on the transaction.
6. Confirm that you can still add a payment method to a customer account or pay for an order through the Order Pay page without any issues.
7. Make a wish for lucky step number 7.



-------------------
